renderdoc/driver/ihv/intel/intel_counters.cpp

```diff
@@ -0,0 +1,423 @@
+        counterDesc.resultType = CompType::Double;
+        switch(metricParams->ResultType)
+        {
+          case RESULT_UINT32:
+            counterDesc.resultType = CompType::UInt;
+            counterDesc.resultByteWidth = sizeof(uint32_t);
+            break;
+          case RESULT_UINT64:
+            counterDesc.resultType = CompType::UInt;
+            counterDesc.resultByteWidth = sizeof(uint64_t);
+            break;
+          case RESULT_FLOAT:
+            counterDesc.resultType = CompType::Float;
+            counterDesc.resultByteWidth = sizeof(float);
+            break;
+          default: break;
+        }
+
+        counterDesc.unit = CounterUnit::Absolute;
+        if(strcmp(metricParams->MetricResultUnits, "cycles") == 0)
+          counterDesc.unit = CounterUnit::Cycles;
+        else if(strcmp(metricParams->MetricResultUnits, "bytes") == 0)
+          counterDesc.unit = CounterUnit::Bytes;
+        else if(strcmp(metricParams->MetricResultUnits, "percent") == 0)
+          counterDesc.unit = CounterUnit::Percentage;
+        else if(strcmp(metricParams->MetricResultUnits, "ns") == 0)
+          counterDesc.unit = CounterUnit::Seconds;
+
+        uint32_t counterId = (uint32_t)GPUCounter::FirstIntel + (uint32_t)m_Counters.size();
+        counterDesc.counter = (GPUCounter)counterId;
+        m_Counters.push_back(counterDesc);
+        addedMetrics.insert(counterDesc.name);
+        m_metricLocation[(GPUCounter)counterId] = std::pair<uint32_t, uint32_t>(j, k);
+        m_counterIds.push_back((GPUCounter)counterId);
+      }
+    }
+  }
+
+  return m_Counters;
+}
+
+uint32_t IntelCounters::GetPassCount()
+{
+  return (uint32_t)m_subscribedMetricSets.size();
+}
+
+uint32_t IntelCounters::BeginSession()
+{
+  if(!m_metricsDevice)
+    return 0;
+
+  m_passIndex = 0;
+  TMetricsDeviceParams_1_2 *deviceParams = m_metricsDevice->GetParams();
+  if(deviceParams->Version.MajorNumber < 1 ||
+     (deviceParams->Version.MajorNumber == 1 && deviceParams->Version.MinorNumber < 2))
+    return 0;
+
+  IOverride_1_2 *frequencyOverride = m_metricsDevice->GetOverrideByName("FrequencyOverride");
+  if(!frequencyOverride)
+    return 0;
+
+  TTypedValue_1_0 *maxFreqSymbol =
+      m_metricsDevice->GetGlobalSymbolValueByName("GpuMaxFrequencyMHz");
+  if(!maxFreqSymbol)
+    return 0;
+
+  TSetFrequencyOverrideParams_1_2 params;
+  params.Enable = true;
+  params.FrequencyMhz = maxFreqSymbol->ValueUInt32;
+  params.Pid = 0;
+  frequencyOverride->SetOverride(&params, sizeof(params));
+  Sleep(500);
+
+  return 0;
+}
+
+void IntelCounters::EndSesssion()
+{
+  if(!m_metricsDevice)
+    return;
+
+  TMetricsDeviceParams_1_2 *deviceParams = m_metricsDevice->GetParams();
+  if(deviceParams->Version.MajorNumber > 1 ||
+     (deviceParams->Version.MajorNumber == 1 && deviceParams->Version.MinorNumber >= 2))
+  {
+    IOverride_1_2 *frequencyOverride = m_metricsDevice->GetOverrideByName("FrequencyOverride");
+
+    if(frequencyOverride)
```

Comment:
Is there any case where this `frequencyOverride` could fail to fetch here, when it fetched successfully above? i.e. is there any risk of the frequency being set to max and then not properly reduced again? Are we allowed to cache the `IOverride_1_2 *` or is that pointless?

Reply:
Frequency override availability depends on platform only. If frequency override was fetched on BeginSession, it also would be resolved on EndSession. We are allowed to cache it but it doesn't make sense.

Reply:
Makes sense, thanks.